sw/host/spiflash/ftdi_spi_interface.cc

```diff
@@ -106,22 +109,82 @@ bool FtdiSpiInterface::TransmitFrame(const uint8_t *tx, uint8_t *rx,
     std::cerr << "Unable to start spi transaction." << std::endl;
     return false;
   }
+
   uint8_t *tmp_rx = ::Transfer(spi_->ctx, tx_local.data(), size);
+  // We're not using the result of this read, so free it right away.
+  if (tmp_rx) {
+    free(tmp_rx);
+  }
+
   if (Stop(spi_->ctx)) {
     std::cerr << "Unable to terminate spi transaction." << std::endl;
-    if (tmp_rx) {
-      free(tmp_rx);
-    }
     return false;
   }
   if (tmp_rx == nullptr) {
     std::cerr << "Unable to transfer data. Frame size: " << size << std::endl;
     return false;
   }
-  usleep(kTransmitDelay);
-  memcpy(rx, tmp_rx, size);
-  free(tmp_rx);
   return true;
 }
+
+bool FtdiSpiInterface::CheckHash(const uint8_t *tx, size_t size) {
+  uint8_t hash[SHA256_DIGEST_LENGTH];
+  SHA256_CTX sha256;
+  SHA256_Init(&sha256);
+  SHA256_Update(&sha256, tx, size);
+  SHA256_Final(hash, &sha256);
+
+  uint8_t *rx;
+
+  int hash_index = 0;
+  bool hash_correct = false;
+
+  if (Start(spi_->ctx)) {
+    std::cerr << "Unable to start spi transaction." << std::endl;
+    return false;
+  }
+
+  auto begin = std::chrono::steady_clock::now();
+  auto now = begin;
+  while (!hash_correct &&
+         std::chrono::duration_cast<std::chrono::microseconds>(now - begin)
+                 .count() < kHashReadTimeoutNs) {
+    usleep(kHashReadDelayNs);
+    rx = nullptr;
+    rx = ::Read(spi_->ctx, size);
```

Comment:
nit: Given things are defined, `size` needs to always be called with the same frame size once initialized. Perhaps this could be parameter of the class interface, initialized at construction time.

Marking as nit since this is something that can be refactored later on.